### ISSUE TITLE
Upstream work on the discoverability of certain recommended volume type aspects.

### ISSUE THREAD

**Issue by josephineSei** [SovereignCloudStack/standards#449]
As a user, I want to be able to look into a volume type object and see all aspects it fulfills so that I can choose the best suited volume type for my volumes.

In #265 a standard for volume types is created. Right now SCS, providers and consumers need to rely on Tags in the description of a volume type to discover volume types with the recommended aspects encrypted and replicated. 

We want to find a way to use the internal extra_spec in volume types for the description of those two aspects, when they are present in a volume type. If this is not possible, we would like to introduce another property, which has to be set by the admin, when setting the volume type. Only after that we will have the possibility to automatically check for a volume type with replication or encyption.

## Definition of Done:

- [ ] The aspect of encryption is discoverable for a user role in upstream OpenStack
- [ ] The aspect of replication is discoverable for a user role in upstream OpenStack
- [ ] The standard is changed to use the upstream way of discovering replication and encryption in a volume type

**Comment by josephineSei** [comment]
For volume types there exist already the concept of user visible extra specs, that are triggered by policy:
https://docs.openstack.org/cinder/latest/admin/user-visible-extra-specs.html
It should be possible to add support for an encryption extra spec, as everything necessary is already there. The encryption parameters can be accessed as an admin like this:
```
stack@woc15:~/devstack$ openstack volume type show LUKS --encryption-type
+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Field              | Value                                                                                                                                                                                                                             |
+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| access_project_ids | None                                                                                                                                                                                                                              |
| description        | None                                                                                                                                                                                                                              |
| encryption         | cipher='aes-xts-plain64', control_location='front-end', created_at='2023-11-30T16:13:15.000000', deleted='False', deleted_at=, encryption_id='872a9738-34d0-43d6-977a-91ea4858e74f', key_size='256', provider='luks', updated_at= |
| id                 | bee35d94-d42c-42d8-b226-32be9ca5b694                                                                                                                                                                                              |
| is_public          | True                                                                                                                                                                                                                              |
| name               | LUKS                                                                                                                                                                                                                              |
| properties         |                                                                                                                                                                                                                                   |
| qos_specs_id       | None                                                                                                                                                                                                                              |
+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
For replication it is harder, as it depends on the used backend. That means, it cannot be derived from inside openstack but would need the input from the provider, whether there is replication or not. I will go into discussion with the Cinder developers about this.

But it might be possible to easily do that as the setting of extra specs is only allowed for admins now and only when the volume type is not in use. So it would just be necessary to make something like this visible to users:
```
openstack volume type set --property replications=3 LUKS
openstack volume type show LUKS --encryption-type
+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Field              | Value                                                                                                                                                                                                                             |
+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| access_project_ids | None                                                                                                                                                                                                                              |
| description        | None                                                                                                                                                                                                                              |
| encryption         | cipher='aes-xts-plain64', control_location='front-end', created_at='2023-11-30T16:13:15.000000', deleted='False', deleted_at=, encryption_id='872a9738-34d0-43d6-977a-91ea4858e74f', key_size='256', provider='luks', updated_at= |
| id                 | bee35d94-d42c-42d8-b226-32be9ca5b694                                                                                                                                                                                              |
| is_public          | True                                                                                                                                                                                                                              |
| name               | LUKS                                                                                                                                                                                                                              |
| properties         | replications='3'                                                                                                                                                                                                                  |
| qos_specs_id       | None                                                                                                                                                                                                                              |
+--------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

This is currently only visible for admins.

**Comment by josephineSei** [comment]
While going through the cinder code to check the places for any upstream patches, I stumbled across the policies:
https://github.com/openstack/cinder/blob/master/cinder/policies/volume_type.py#L156
This allows users to read the extra specs but it can be shut down through a policy file (only allowing admin access). We should keep that in mind.

**Comment by josephineSei** [comment]
I am currently trying to implement a visible user extra spec for encryption, that is automatically set when creating or unset when deleting an encryption type in here:
https://github.com/openstack/cinder/blob/master/cinder/api/contrib/volume_type_encryption.py
It should not be setable outside of that workflow.

**Comment by josephineSei** [comment]
I am currently running into some errors with my implementation and try to fix that:
```
openstack volume type create --encryption-provider nova.volume.encryptors.luks.LuksEncryptor --encryption-cipher aes-xts-plain64 --encryption-key-size 256 --encryption-control-location front-end LUKS
Failed to set encryption information for this volume type: The resource could not be found.<br /><br />


 (HTTP 404)
local variable 'encryption' referenced before assignment
```

**Comment by josephineSei** [comment]
I pushed my changes to https://review.opendev.org/c/openstack/cinder/+/907519 to discuss them with upstream.

**Comment by josephineSei** [comment]
> I pushed my changes to https://review.opendev.org/c/openstack/cinder/+/907519 to discuss them with upstream.

I was fixing a few Errors and found out, that creating custom properties / extra specs lets the volume creation fail. I am asking Upstream about the details of this.

**Comment by josephineSei** [comment]
I am investigating a hint that it might be the scheduler, who is responsible for the errors of custom properties. Mailing List Discussion: https://lists.openstack.org/archives/list/openstack-discuss@lists.openstack.org/thread/ZWXFYIQ3FSFC5MGTSIMVEHCEJRQQRQ3X/

**Comment by josephineSei** [comment]
> I am investigating a hint that it might be the scheduler, who is responsible for the errors of custom properties. Mailing List Discussion: https://lists.openstack.org/archives/list/openstack-discuss@lists.openstack.org/thread/ZWXFYIQ3FSFC5MGTSIMVEHCEJRQQRQ3X/

The cause of this is indeed the Cinder Scheduler, that checks ALL properties (== extra_specs) here, for their compatability with the backends: https://github.com/openstack/cinder/blob/master/cinder/scheduler/filters/capabilities_filter.py#L56 

Changing this Filter would be possible, but whether we would only allow one or two special extra_specs to be skipped in this filter or to cheange the filter using a list of extra_specs to check for their presence and compatability is something we need to discuss with upstream.

**Comment by josephineSei** [comment]
I attended the Cinder meeting to discuss my findings, but unfortunately there was not enough time, we will continue the discussion via ML or in the next week.

https://meetings.opendev.org/meetings/cinder/2024/cinder.2024-02-07-14.00.log.html

**Comment by fkr** [comment]
thanks @josephineSei for keeping this updated.

**Comment by josephineSei** [comment]
After trying to reach out to people and going through the patches again, I added this as a discussion point for the midcycle ptg: https://etherpad.opendev.org/p/cinder-caracal-midcycles

**Comment by josephineSei** [comment]
To be better prepared for the discussion tomorrow, I try to look for alternative ways to include the visibility of either volume type defined or administrator defined aspects in volume types.

While it may be possible to include a method in the API call for showing a volume type, that gives normal users information about whether an encryption type is present or not, this is not a feasible way for the replication, that is backend-specific. This aspect has to be set by the admin explicitly but is NOT allowed to interact with the volume scheduler, as it might lead to Errors.

```
mysql> describe volume_types;
+--------------+--------------+------+-----+---------+-------+
| Field        | Type         | Null | Key | Default | Extra |
+--------------+--------------+------+-----+---------+-------+
| created_at   | datetime     | YES  |     | NULL    |       |
| updated_at   | datetime     | YES  |     | NULL    |       |
| deleted_at   | datetime     | YES  |     | NULL    |       |
| deleted      | tinyint(1)   | YES  |     | NULL    |       |
| id           | varchar(36)  | NO   | PRI | NULL    |       |
| name         | varchar(255) | YES  |     | NULL    |       |
| qos_specs_id | varchar(36)  | YES  | MUL | NULL    |       |
| is_public    | tinyint(1)   | YES  |     | NULL    |       |
| description  | varchar(255) | YES  |     | NULL    |       |
+--------------+--------------+------+-----+---------+-------+
9 rows in set (0.01 sec)

mysql> describe volume_type_extra_specs;
+----------------+--------------+------+-----+---------+----------------+
| Field          | Type         | Null | Key | Default | Extra          |
+----------------+--------------+------+-----+---------+----------------+
| created_at     | datetime     | YES  |     | NULL    |                |
| updated_at     | datetime     | YES  |     | NULL    |                |
| deleted_at     | datetime     | YES  |     | NULL    |                |
| deleted        | tinyint(1)   | YES  |     | NULL    |                |
| id             | int          | NO   | PRI | NULL    | auto_increment |
| volume_type_id | varchar(36)  | NO   | MUL | NULL    |                |
| key            | varchar(255) | YES  |     | NULL    |                |
| value          | varchar(255) | YES  |     | NULL    |                |
+----------------+--------------+------+-----+---------+----------------+
8 rows in set (0.00 sec)

mysql> describe encryption;
+------------------+--------------+------+-----+---------+-------+
| Field            | Type         | Null | Key | Default | Extra |
+------------------+--------------+------+-----+---------+-------+
| created_at       | datetime     | YES  |     | NULL    |       |
| updated_at       | datetime     | YES  |     | NULL    |       |
| deleted_at       | datetime     | YES  |     | NULL    |       |
| deleted          | tinyint(1)   | YES  |     | NULL    |       |
| cipher           | varchar(255) | YES  |     | NULL    |       |
| control_location | varchar(255) | NO   |     | NULL    |       |
| key_size         | int          | YES  |     | NULL    |       |
| provider         | varchar(255) | NO   |     | NULL    |       |
| volume_type_id   | varchar(36)  | NO   |     | NULL    |       |
| encryption_id    | varchar(36)  | NO   | PRI | NULL    |       |
+------------------+--------------+------+-----+---------+-------+
10 rows in set (0.01 sec)
```
The database shows, that the encryption parameters are saved in a different table, while something to simply store information would be the extra_specs table or to create a new additional table. This would have the downside, that next to the extra_specs table there would be a need for an additional `metadata` table or something alike. This would blow up the database and it would put work on the API methods and maybe even the view of the volume type objects. In that case the extra specs and metadata would share the "properties field" of the volume type object:

```
openstack volume type show LUKS --encryption-type
+--------------------+----------------------------------------------------------------------+
| Field              | Value                                                                |
+--------------------+----------------------------------------------------------------------+
| access_project_ids | None                                                                 |
| description        | None                                                                 |
| encryption         | cipher='aes-xts-plain64', control_location='front-end',              |
|                    | created_at='2024-02-01T09:33:28.000000', deleted='False',            |
|                    | deleted_at=, encryption_id='d7995c8e-83a4-4535-94ee-151e482dac2e',   |
|                    | key_size='256',                                                      |
|                    | provider='nova.volume.encryptors.luks.LuksEncryptor', updated_at=    |
| id                 | 39bc174e-c075-4296-8a09-f3b274ae0caa                                 |
| is_public          | True                                                                 |
| name               | LUKS                                                                 |
| properties         |                                                                      |
| qos_specs_id       | None                                                                 |
+--------------------+----------------------------------------------------------------------+
```

A better option would be to put the extra_specs into a separate field and leave the properties field for other metadata.
But I doubt that it would be likely to adjust OpenStack like this.

- In the end I can conclude, that the best option would be to open up the extra_specs to also hold key-value pairs that are not used by the scheduler. (most likely the easiest way to implement - no API or DB changes)
- The second best option would be to define a new field for any user input and information, that works exactly like the properties field and would require a new DB table, new API endpoints and some rework for the views. (API and DB changes, but backwards compatible)
- The third option would be to put the information into the `properties` field mixed with the extra_specs but only in the API. Having a separation between two tables in the DB and just mixing those inputs, when information is requested. (new DB table, API behavior might change, a lot of work in separating the properties -> imho this is Error prone)
- The last option would be to define a new field and move all extra_specs into that field and keep the properties field for any other user visible information. (DB and API changes, NOT backward compatible, best separation but most unlikely option)

**Comment by josephineSei** [comment]
I have attended the Cinder midcycle meeting to discuss about the use cases we have and what we want to achieve. I explained everything and we also discussed about several options to achieve this (like the ones I mentioned in the last comment)
Unfortunately they also have no certain direction to push me to. So the discussion on how we can achieve to have it will continue on the ML and may even result in a spec for Cinder. Most likely we either only get an easy way to get the ecnryption OR we might need to adjust DB and API with a new metadata field.

The etherpad of the midcycle with an Action Item for me: https://etherpad.opendev.org/p/cinder-caracal-midcycles

**Comment by josephineSei** [comment]
I have created a blueprint[^1] and am working out a spec for the change in the volume types (still under work, as I also have to consider all DB, API and other impacts ALL options would have right now). I wrote an email to the mailinglist[^2] already containing the options to discuss, so the topic gets more attention.

[^1]: https://blueprints.launchpad.net/cinder/+spec/user-visible-information-in-volume-types
[^2]: https://lists.openstack.org/archives/list/openstack-discuss@lists.openstack.org/thread/PP7IMXVOO2SM47JRMDYYVB2IA3XIEZD5/

**Comment by josephineSei** [comment]
I attended the Public Cloud Sig to align with the process of discoverability in overall openstack.

**Comment by josephineSei** [comment]
I edited the spec according to the review I got and fixed some grammatical errors.

**Comment by josephineSei** [comment]
I prepared the spec patch for the next cycle.

**Comment by josephineSei** [comment]
After having the new document dturcture for the next cycle I updated the spec patch again.
https://review.opendev.org/c/openstack/cinder-specs/+/909195

**Comment by josephineSei** [comment]
I prepared myself for the dicussion of the different ways for user visible information in volume types in the PTG that will happen today or maybe tomorrow.

**Comment by josephineSei** [comment]
### Results of the PTG discussion

I presented some options on how to tackle the user visibility and we came to an agreement:

1. A new DB table for volume_type_metadata should be exclusively used for information from the deployer for the user.
2. This should be presented in a new metadata field for volume types in the API.
3. To standardize the phrases in the metadata field, metadefs can be used in addition.

### Metadef API

The metadef API currently does exist in Glance:
https://docs.openstack.org/api-ref/image/v2/metadefs-index.html

And from this document it will define metadata keys for various resources in OpenStack:
https://docs.openstack.org/glance/latest/user/metadefs-concepts.html

Adding a new definition there seems to be the equivalent to a standard in scs. Users could apply to this but are not forced to.

Nevertheless for SCS it might be worth looking into this metadefs API in a separate issue.

For the scope of this issue this is just a possible second phase after introducing a way for deployers to add metadata to volume types.

### Work items

- [ ] rewrite the spec
- [ ] add patches to Cinder after the spec was merged (only for the DB and the API)
- [ ] add patches to OSC/SDK to be able to add something to the new metadata field

- [ ] adjust Standard and Test Script to use the new field (maybe done after the dalmatian release)

2nd Phase:

- [ ] add metadef description for encrypted and replicated volume types

**Comment by josephineSei** [comment]
I updated the spec: https://review.opendev.org/c/openstack/cinder-specs/+/909195

**Comment by josephineSei** [comment]
To be prepared for the next Cinder meeting, I am updating my devstack node. 
I want to start implementing on the current master.
But I was running into some issues with the new release. One of them was that glance did not answer - it wasn't even rolled out correctly in combination with Ceph. I found the reason randomly and it had nothing to to with the Error message that was: 
```
Failed to contact the endpoint at http://.... for discovery. Fallback to using that endpoint as the base url. 
Failed to contact the endpoint at http://.... for discovery. Fallback to using that endpoint as the base url. 
The image service for : exists but does not have any supported versions.
```
It was a configuration in the glance-api.conf . The new default for "image_cache_driver" since 2024.1 is "centralized_db", but something in the glance process expected one of the older config values. After I changed it, I was able to contact the glance endpoint.

**Comment by josephineSei** [comment]
Today I finished the work on my devstack (at least I think so), was re-iterating and presenting the spec at the Cinder team meeting and began resarch on the implementation.

This will be done in small parts:
1. Implementing an upgrade path to add a new db table: volume_type_metadata
2. Adding all database requests for this new table
3. Adjusting the current API to also show the new metadata field
4. Adding the new API to set and unset new metadata.

**Comment by josephineSei** [comment]
I am implementing the first part of upgrading the Cinder DB. currently my devstack seems to not run through all cinder migrations of the DB.

**Comment by josephineSei** [comment]
I added a first patchset with the database upgrade: https://review.opendev.org/c/openstack/cinder/+/918316
I am working on the API extension to see the metadata in the volume type object. This will also be put in that patch set above.

**Comment by josephineSei** [comment]
Having part 1 in the first patchset, I started working on part three to actually show the metadata field in the volume type.

**Comment by josephineSei** [comment]
I was looking through the open review comments on the spec

**Comment by josephineSei** [comment]
Adjusted the spec another time

**Comment by josephineSei** [comment]
I restarted working on the patchsets for this, but currently I am fighting a bug, which gives me only an error when asking for the volume_types via CLI.

**Comment by josephineSei** [comment]
With the spec freeze dooming, I polished the spec again and asked for more reviews, because the spec is in good standing and might only need workflow right now to proceed.

**Comment by josephineSei** [comment]
It seems that the model I was adding in with the first  is not enough, because when sqlalchemy tries to load the table it does this for the extra specs:
```
extra_specs model: VolumeType.extra_specs
```
But this for my metadata model:
```
metadata model: MetaData()
```

I am currently digging through the Cinder part to see, where this could be changed.

**Comment by josephineSei** [comment]
I found the reason. Now I do have the metadata in the response for list and show volume type:
```
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| access_project_ids | None                                 |
| description        | None                                 |
| id                 | 15002270-b9ed-4b7f-ba08-cea35f3acc1f |
| is_public          | True                                 |
| metadata           | None                                 |
| name               | LUKS                                 |
| properties         |                                      |
| qos_specs_id       | None                                 |
+--------------------+--------------------------------------+
```

The list volume type currently does not show the metadata, this is handled in the OpenStack Client - so I would need to change there:
```
RESP BODY: {"volume_types": [{"id": "15002270-b9ed-4b7f-ba08-cea35f3acc1f", "name": "LUKS", "is_public": true, "description": null, "metadata": null, "extra_specs": {}, "qos_specs_id": null, "os-volume-type-access:is_public": true}, {"id": "74241c6f-6678-4388-9f69-d16d91793cc1", "name": "ceph", "is_public": true, "description": null, "metadata": null, "extra_specs": {"volume_backend_name": "ceph"}, "qos_specs_id": null, "os-volume-type-access:is_public": true}, {"id": "41a4e540-0660-4562-aeb1-e3891121c88f", "name": "__DEFAULT__", "is_public": true, "description": "Default Volume Type", "metadata": null, "extra_specs": {}, "qos_specs_id": null, "os-volume-type-access:is_public": true}]}
GET call to volumev3 for http://192.168.23.161/volume/v3/db05e1116cb04ac689d82c738a5f103c/types?is_public=None used request id req-5df468ab-5105-44e6-abee-509dadab0d48
+--------------------------------------+-------------+-----------+---------------------+----------------------------+
| ID                                   | Name        | Is Public | Description         | Properties                 |
+--------------------------------------+-------------+-----------+---------------------+----------------------------+
| 15002270-b9ed-4b7f-ba08-cea35f3acc1f | LUKS        | True      | None                |                            |
| 74241c6f-6678-4388-9f69-d16d91793cc1 | ceph        | True      | None                | volume_backend_name='ceph' |
| 41a4e540-0660-4562-aeb1-e3891121c88f | __DEFAULT__ | True      | Default Volume Type |                            |
+--------------------------------------+-------------+-----------+---------------------+----------------------------+
```
I may need to restructure the metadata to a dict object. But currently am working on the unit tests.

**Comment by josephineSei** [comment]
I worked on the failing unit tests and transfomring the metadata to a dict after getting it from the database:
https://review.opendev.org/c/openstack/cinder/+/918316

**Comment by josephineSei** [comment]
The spec was merged: https://review.opendev.org/c/openstack/cinder-specs/+/909195
I  rebased my patchset and started working on the unit tests again:
https://review.opendev.org/c/openstack/cinder/+/918316

**Comment by josephineSei** [comment]
The patchset right now includes the adding of a db table and the handling of showing the new metadata within the lisiting and showing of volume types. (All within the API). The unit tests for this are working.

What is still missing:
1. I need to work on the workflow of adding metadata key-value pairs to the API (and then database)
2. I will take a look into the tempest tests (need adjustments)

**Comment by josephineSei** [comment]
I already identified something that need adjustment in tempest and currently prepare my devstack to work with tempest.

**Comment by josephineSei** [comment]
I started looking into the patches again and started working on the API, when I realized, that Feature Freeze in OpenStack will be next week.
I don't think, that these patches will be part of the next release, because parts are still missing and need implementation and there was no review until now.
I will concentrate on the image encryption first, and come back here, when I have time.

**Comment by josephineSei** [comment]
After having the image encryption patch ready for review, I resumed work on the API endpoint today.

**Comment by josephineSei** [comment]
I paused the work on this again, to help with the image encryption #560 .

**Comment by josephineSei** [comment]
I started patching the openstack-client,  for easier testing: https://review.opendev.org/c/openstack/python-openstackclient/+/928132

While testing without the functional API-part (the patched API-endpoint - everything else: database, write and read is there) I was expecting a failure and some Error message.
But I got this:
```
(venv) stack@devstack:~/devstack$ openstack volume type show test
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| access_project_ids | None                                 |
| description        | None                                 |
| id                 | 87f89839-6f55-4548-b041-e2b451dbcf89 |
| is_public          | True                                 |
| metadata           | {'cat': 'miau'}                      |
| name               | test                                 |
| properties         |                                      |
| qos_specs_id       | None                                 |
+--------------------+--------------------------------------+
(venv) stack@devstack:~/devstack$ openstack volume type set --metadata "tiger"="roar" test
(venv) stack@devstack:~/devstack$ openstack volume type show test
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| access_project_ids | None                                 |
| description        | None                                 |
| id                 | 87f89839-6f55-4548-b041-e2b451dbcf89 |
| is_public          | True                                 |
| metadata           | {'cat': 'miau'}                      |
| name               | test                                 |
| properties         | tiger='roar'                         |
| qos_specs_id       | None                                 |
+--------------------+--------------------------------------+
```

That means, somewhere within either the clients or the API the `extra_specs` are not only referred to as `properties` but also as `metadata`. I will need to find the place and then have two options: 
1. looking through the code to use `metadata` not for the `extra_specs`
2. renaming my volume type `metadata` to something else

I will also need to discuss this with the Cinder team, but I think they will go for the second option, as someone could right now make use of the fact that `metadata` ends up to be `extra_specs` already.

**Comment by josephineSei** [comment]
I aligned the Openstack client and cinder client patch (https://review.opendev.org/c/openstack/python-cinderclient/+/928352) and am currently working on the cinder API - I need a new API endpoint, because there is also a new endpoint for the extra-specs and we need to differentiate between the metadata and the extra_specs/properties. PLUS: I don't think mixing those two things up would be something the Cinder team agrees with.

**Comment by josephineSei** [comment]
The new API endpoint needs more work, but the clients already call it correctly:

```
http://192.168.23.161:80 "POST /volume/v3/db05e1116cb04ac689d82c738a5f103c/types/87f89839-6f55-4548-b041-e2b451dbcf89/metadata HTTP/1.1" 404 112
RESP: [404] Connection: close Content-Length: 112 Content-Type: application/json Date: Mon, 09 Sep 2024 12:26:15 GMT Server: Apache/2.4.52 (Ubuntu) x-openstack-request-id: req-b51c6a12-6790-4288-b77c-832ddb10a6b1
```

There are a lot of small adjustments and I think I will add another patch for the whole API stuff. This is better for reviewing.

**Comment by josephineSei** [comment]
The work in progress patch for the API endpoint for the volume type metadata: https://review.opendev.org/c/openstack/cinder/+/928794
I still need to somehow register this API extension, currently the endpoint is not found.

**Comment by josephineSei** [comment]
After testing and fixing minor bugs, I was able to use my patches to create a new metadata for a volume type from the CLI:
```
stack@devstack:~/devstack$ openstack volume type show test
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| access_project_ids | None                                 |
| description        | None                                 |
| id                 | 87f89839-6f55-4548-b041-e2b451dbcf89 |
| is_public          | True                                 |
| metadata           | {'cat': 'miau'}                      |
| name               | test                                 |
| properties         |                                      |
| qos_specs_id       | None                                 |
+--------------------+--------------------------------------+
stack@devstack:~/devstack$ openstack volume type set --metadata "tiger"="roar" test
stack@devstack:~/devstack$ openstack volume type show test
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| access_project_ids | None                                 |
| description        | None                                 |
| id                 | 87f89839-6f55-4548-b041-e2b451dbcf89 |
| is_public          | True                                 |
| metadata           | {'tiger': 'roar', 'cat': 'miau'}     |
| name               | test                                 |
| properties         |                                      |
| qos_specs_id       | None                                 |
+--------------------+--------------------------------------+
```

I need to also check the unset, and the workflow for updating.

**Comment by josephineSei** [comment]
Updating a volume type metadata works:
```
stack@devstack:~/devstack$ openstack volume type show test
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| access_project_ids | None                                 |
| description        | None                                 |
| id                 | 87f89839-6f55-4548-b041-e2b451dbcf89 |
| is_public          | True                                 |
| metadata           | {'tiger': 'roar', 'cat': 'miau'}     |
| name               | test                                 |
| properties         |                                      |
| qos_specs_id       | None                                 |
+--------------------+--------------------------------------+
stack@devstack:~/devstack$ openstack volume type set --metadata "tiger"="growl" test
stack@devstack:~/devstack$ openstack volume type show test
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| access_project_ids | None                                 |
| description        | None                                 |
| id                 | 87f89839-6f55-4548-b041-e2b451dbcf89 |
| is_public          | True                                 |
| metadata           | {'cat': 'miau', 'tiger': 'growl'}    |
| name               | test                                 |
| properties         |                                      |
| qos_specs_id       | None                                 |
+--------------------+--------------------------------------+
```

Deleting a metadata still got some bugs in it.

**Comment by josephineSei** [comment]
```
stack@devstack:~/devstack$ openstack volume type unset --metadata tiger test
stack@devstack:~/devstack$ openstack volume type show test
+--------------------+--------------------------------------+
| Field              | Value                                |
+--------------------+--------------------------------------+
| access_project_ids | None                                 |
| description        | None                                 |
| id                 | 87f89839-6f55-4548-b041-e2b451dbcf89 |
| is_public          | True                                 |
| metadata           | {'cat': 'miau'}                      |
| name               | test                                 |
| properties         |                                      |
| qos_specs_id       | None                                 |
+--------------------+--------------------------------------+
```

The things still missing for functionality are: creating a volume type with some metadata
Another thing is the ability to filter by volume type metadata - this is not yet done.

For upstream this is too late for 2024.2, but can be targeted for the next cycle. There are tests missing like unit tests and tempest tests. These are also needed for upstream.

**Comment by josephineSei** [comment]
I started looking into implementing a filter. When listing all volume types I want to be able to filter by the presence of certain metadata.

**Comment by josephineSei** [comment]
I started taking a look into the failing tests and will also add unit tests.

**Comment by josephineSei** [comment]
I rebased the patches and added unit tests for the cinderclient. I will continue with the other patches.

**Comment by josephineSei** [comment]
The pipeline for the client patches is green.
I am now investigating the Errors on the Cinder patches, to prepare for the next PTG.

**Comment by josephineSei** [comment]
The patch, which adds the database table (https://review.opendev.org/c/openstack/cinder/+/918316) and just also ask for the metadata table content, when the volume types are listed / showed - has only one failing unit test, which is not due to the changes, some mypy failures (also in code parts I did not change) and tempest tests, which require the tempest change to be in place.

The first things cannot be handled by me, and the tempest tests - i looked through the failing tests and it seems that all (also the failing plugin tests) seem to fail validation, which needs the schema adjustment in tempest. So there are two patches which rely on each other, that cannot be merged without each other - I will ask Cinder, how they deal with such things.

Also I think there is no more unit tests needed for this patch then the ones I adjusted and the one I wrote, that checks the db upgrade itself.

I now start looking into the other Cinder patch, that focuses on adding new API endpoints, and the handling between endpoints and the db.

**Comment by josephineSei** [comment]
While implementing the missing shown metadata column in the `openstack volume type list --long` answer and trying to fix the unit tests therefore, I once again came to the point to think about microversions. Cinder requires microversions for each new API endpoint / changes response, etc... So the Cinder API part will definitely need a microversion and therefor also the openstackclient and the unit tests in the openstack client have to be aware of that microversion.

I will hold on further implementing unit tests for the client and start digging into the microversion stuff more. So this will not yet be commited to upstream patches:
```
stack@devstack:~/devstack$ openstack volume type list --long
+--------------------------------------+-------------+-----------+--------------------------------------------------------------------------+------------+
| ID                                   | Name        | Is Public | Description                                                              | Properties |
+--------------------------------------+-------------+-----------+--------------------------------------------------------------------------+------------+
| 87f89839-6f55-4548-b041-e2b451dbcf89 | test        | True      | None                                                                     |            |
| 15002270-b9ed-4b7f-ba08-cea35f3acc1f | LUKS        | True      | [scs:encrypted, replicated] Content will be replicated three times to    |            |
|                                      |             |           | ensure consistency and availability for your data. LUKS encryption is    |            |
|                                      |             |           | used.                                                                    |            |
| 74241c6f-6678-4388-9f69-d16d91793cc1 | ceph        | True      | None                                                                     |            |
| 41a4e540-0660-4562-aeb1-e3891121c88f | __DEFAULT__ | True      | Default Volume Type                                                      |            |
+--------------------------------------+-------------+-----------+--------------------------------------------------------------------------+------------+
clean_up ListVolumeType: 
END return value: 0
stack@devstack:~/devstack$ nano /opt/stack/data/venv/lib/python3.10/site-packages/openstackclient/volume/v3/volume_type.py
stack@devstack:~/devstack$ openstack volume type list --long
+--------------------------------------+-------------+-----------+------------------------------------------------------------+------------+---------------------------+
| ID                                   | Name        | Is Public | Description                                                | Properties | Metadata                  |
+--------------------------------------+-------------+-----------+------------------------------------------------------------+------------+---------------------------+
| 87f89839-6f55-4548-b041-e2b451dbcf89 | test        | True      | None                                                       |            | cat='miau', tiger='growl' |
| 15002270-b9ed-4b7f-ba08-cea35f3acc1f | LUKS        | True      | [scs:encrypted, replicated] Content will be replicated     |            |                           |
|                                      |             |           | three times to ensure consistency and availability for     |            |                           |
|                                      |             |           | your data. LUKS encryption is used.                        |            |                           |
| 74241c6f-6678-4388-9f69-d16d91793cc1 | ceph        | True      | None                                                       |            |                           |
| 41a4e540-0660-4562-aeb1-e3891121c88f | __DEFAULT__ | True      | Default Volume Type                                        |            |                           |
+--------------------------------------+-------------+-----------+------------------------------------------------------------+------------+---------------------------+
```

**Comment by josephineSei** [comment]
I started adding a microversion for my patch by following this documentation: /home/kiari/Dokumente/upstream/cinder/api-ref/source/v3/samples/volume_type/volume-type-show-response.json
I am working on point 8 of the "other necessary changes" before I will add the decorators. I pushed my current state of work to this patch: https://review.opendev.org/c/openstack/cinder/+/928794

**Comment by josephineSei** [comment]
After introducing the microversion I was able to simplify the first patch to add the database table:
https://review.opendev.org/c/openstack/cinder/+/918316

This is good, because smaller changes are easier to review :)

**Comment by josephineSei** [comment]
I looked through the failing tests in the database and the API patch, most of them seem to come from other problems. I made a few adjustments to the API patch to fix the failing ones, my patch was responsible for.
I now start adding unit tests for the new API version, hopefully having some ready for the PTG this week

**Comment by josephineSei** [comment]
I am working on unit tests but can only find small amounts of time due to the PTG and other stuff.

**Comment by josephineSei** [comment]
We discussed a few points at the PTG and the Cinder teams promised me reviews. I already started implementing some feedback - like kicking out notifications.
Another point is, that I somehow need to test the performance of the db calls, because, with many types and and request the db calls might have a bigger impact on performance. I need to look through this and whether I can test this on my devstack.

**Comment by josephineSei** [comment]
Back on working on the unit test and thinking about the db performance

**Comment by josephineSei** [comment]
I marked the spec as active and need to investigate a bit more in unit tests, the first test seems to run on my machine: https://review.opendev.org/c/openstack/cinder/+/928794

**Comment by josephineSei** [comment]
I stumbled upon a functional unit test, that has a very strange behavior:

When I execute the functional unit test to check the extensions I get an ERROR:
```
$ tox -efunctional-py39  -vv -- cinder.tests.functional.api_sample_tests.test_extensions.ExtensionsSampleJsonTest.test_extensions
....
....
==============================
Failed 1 tests - output below:
==============================

cinder.tests.functional.api_sample_tests.test_extensions.ExtensionsSampleJsonTest.test_extensions
-------------------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 394, in _verify_response
    response_result = self._compare_result(template_data,

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 280, in _compare_result
    matched_value = self._compare_dict(

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 183, in _compare_dict
    res = self._compare_result(expected[key], result[key],

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 284, in _compare_result
    matched_value = self._compare_list(

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 238, in _compare_list
    raise NoMatch('\n'.join(error))

    cinder.tests.functional.api_samples_test_base.NoMatch: Extra list items in Response:
{'alias': 'os-types-metadata', 'description': 'Type metadata support.', 'links': [], 'name': 'TypesMetadata', 'updated': '2024-09-09T00:00:00+00:00'}
```


When I try to add the new microversion in the extension-list-response sample, I get that Error:
```
==============================
Failed 1 tests - output below:
==============================

cinder.tests.functional.api_sample_tests.test_extensions.ExtensionsSampleJsonTest.test_extensions
-------------------------------------------------------------------------------------------------

Captured traceback:
~~~~~~~~~~~~~~~~~~~
    Traceback (most recent call last):

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 416, in _verify_response
    self._compare_result(template_data, sample_data, "Sample")

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 280, in _compare_result
    matched_value = self._compare_dict(

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 183, in _compare_dict
    res = self._compare_result(expected[key], result[key],

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 284, in _compare_result
    matched_value = self._compare_list(

      File ".../cinder/cinder/tests/functional/api_samples_test_base.py", line 238, in _compare_list
    raise NoMatch('\n'.join(error))

    cinder.tests.functional.api_samples_test_base.NoMatch: Extra list items in template:
{'alias': 'os-types-metadata', 'description': 'Type metadata support.', 'links': [], 'name': 'TypesMetadata', 'updated': '%(extension_update)s'}
```

The failing part here is not in the original test file, but in the abstract test-base class, where three things are compared:
1. The response from the actual call to the function
2. The template I first adjusted
3. a document in the api-ref

This is hard to detect, because the original test does not indicate the not only 2 things are compared, but in fact 3.
This can only be concluded, when going through the base class.

So whenever there is a template to be adjusted, it may not be the only thing to adjust in Cinder so that your unit tests will be successful.

**Comment by josephineSei** [comment]
I am continuing working on unit tests and attended the Cinder weekly meeting to raise attention for reviews.

**Comment by josephineSei** [comment]
unit tests adjustments to reflect the new microversion is going into a good direction: https://review.opendev.org/c/openstack/cinder/+/928794

**Comment by josephineSei** [comment]
Only a few more unit tests to add here: https://review.opendev.org/c/openstack/cinder/+/928794/17/cinder/tests/unit/api/contrib/test_types_metadata.py

**Comment by josephineSei** [comment]
I finished unit tests, I think. I will take another look in the functional and maybe tempest tests.

**Comment by josephineSei** [comment]
After fixing some failures in the OpenStack zuuil pipeline, I started looking into creating tempest tests with a microversion. This leads to the question, whether tests should reside in tempest itself or rather in the tempest plugin from Cinder.
For now I think it should be tempest itself, but it could be argued, that these are just API and database calls and thus should rather reside in the cinder tempest plugin. You can read [here](https://docs.openstack.org/tempest/latest/microversion_testing.html) when tests should be part of tempest and when not.

**Comment by josephineSei** [comment]
I have a first draft of tempest tests, which should be sufficient in the amount of test cases: https://review.opendev.org/c/openstack/tempest/+/935142
I still need to look into the pipeline and check possible errors.

**Comment by josephineSei** [comment]
Momentan funktioniert der Aufruf mit der korrekten microversion in meinen tempest tests nicht. Daher schlagen sie fehl: https://review.opendev.org/c/openstack/tempest/+/935142

Ich versuche das Problem anzugehen und zu lösen.

**Comment by josephineSei** [comment]
I'm still on the tempest tests and also updated the clients to reflect the use of microversions. I still need to test all the patches together.

**Comment by josephineSei** [comment]
The documentation about the usage of microversions in tempest is not very clear. I am still struggeling with this.
And I missed the dependence of the cinderclient patch on the cinder-API patch. I hope that fixes the CI-pipeline.

**Comment by josephineSei** [comment]
I tried a few more things to address that the tests use the correct microversion in the tempest tests.

**Comment by josephineSei** [comment]
The patches are in a good state, I just could not get tempest to accept the microversions:
https://review.opendev.org/q/topic:%22user-visible-metadata-in-volume-types%22

Nevertheless the major part of work is done, it needs reviews from upstream openstack cinder team and hopefully only smaller adjustments.
